torchtune/models/flamingo/_component_builders.py

```diff
@@ -0,0 +1,257 @@
+            the size of the input image. In this case, the function will consider your image as a single tile.
+        max_num_tiles (int): The maximum number of tiles that can be processed. This is used to
+            determine the size of the positional embeddings.
+        in_channels (int): The number of image input channels.
+
+    Returns:
+        FlamingoEncoder: Instantiation of Flamingo encoder.
+    """
+
+    # clip encoder
+    clip = clip_vision_encoder(
+        tile_size=tile_size,
+        patch_size=patch_size,
+        embed_dim=clip_embed_dim,
+        num_layers=clip_num_layers,
+        num_heads=num_heads,
+        out_indices=clip_hidden_states,
+        max_num_tiles=max_num_tiles,
+        in_channels=in_channels,
+        output_cls_projection=False,
+    )
+
+    # Projection head
+    mlp_ratio = 4
+    hidden_dim = int(mlp_ratio * clip_embed_dim)
+    head_dim = clip_embed_dim // num_heads
+    num_kv_heads = num_heads
+
+    self_attn = MultiHeadAttention(
+        embed_dim=clip_embed_dim,
+        num_heads=num_heads,
+        num_kv_heads=num_heads,
+        head_dim=head_dim,
+        q_proj=nn.Linear(clip_embed_dim, num_heads * head_dim, bias=True),
+        k_proj=nn.Linear(clip_embed_dim, num_kv_heads * head_dim, bias=True),
+        v_proj=nn.Linear(clip_embed_dim, num_kv_heads * head_dim, bias=True),
+        output_proj=nn.Linear(clip_embed_dim, clip_embed_dim, bias=True),
+        pos_embeddings=None,
+        attn_dropout=0.0,
+        is_causal=False,
+    )
+
+    mlp = clip_mlp(
+        in_dim=clip_embed_dim,
+        hidden_dim=hidden_dim,
+        out_dim=clip_embed_dim,
+        activation=nn.GELU(),
+    )
+
+    layer = TransformerSelfAttentionLayer(
+        attn=self_attn,
+        mlp=mlp,
+        sa_norm=Fp32LayerNorm(clip_embed_dim, eps=1e-5),
+        mlp_norm=Fp32LayerNorm(clip_embed_dim, eps=1e-5),
+        sa_scale=TanhGate(),
+        mlp_scale=TanhGate(),
+    )
+
+    # we concatenate clip embeddings and hidden layers output
+    # and project it to embed_dim_out, which will be used for the
+    # cross encoding
+    num_hidden_inputs = len(clip_hidden_states) if clip_hidden_states is not None else 0
+    proj_in = clip_embed_dim * (num_hidden_inputs + 1)
+    projection_head = FlamingoProjectionHead(
+        layer=layer,
+        num_layers=num_layers_projection,
+        output=nn.Linear(proj_in, decoder_embed_dim),
+        num_hidden_inputs=num_hidden_inputs
+    )
+
+    return FlamingoEncoder(encoder=clip, projection_head=projection_head)
+
+
+def flamingo_decoder(
+    vocab_size: int,
+    num_layers: int,
+    fusion_interval: int,
+    num_special_tokens: int,
+    num_heads: int,
+    num_kv_heads: int,
+    embed_dim: int,
+    max_seq_len: int,
+    rope_base: int = 500000.0,
+    intermediate_dim: Optional[int] = None,
+) -> TransformerDecoder:
+    """
+    Build the decoder associated with the Llama3 model with additonal fused
```

Comment:
I'm once more slightly confused by the model naming in the docstrings, which is likely due to my lack of knowledge here. This is a builder for a flamingo decoder where the language model component is llama3?

Reply:
Here we're using flamingo to refer to the general architecture and not the specific implementation of the original flamingo model from DeepMind. So to modernize the approach we're using llama3 and CLIP instead. This will allow us to test our deep fusion setup.